packages/asset-swapper/src/utils/aggregate.ts

```diff
@@ -0,0 +1,780 @@
+ * Represents an individual DEX sample from the sampler contract.
+ */
+export interface DexSample {
+    source: ERC20BridgeSource;
+    input: BigNumber;
+    output: BigNumber;
+}
+
+/**
+ * Flags for `Fill` objects.
+ */
+export enum FillFlags {
+    SourceNative = 0x1,
+    SourceUniswap = 0x2,
+    SourceEth2Dai = 0x4,
+    SourceKyber = 0x8,
+}
+
+/**
+ * Improve a market sell operation by (potentially) merging native orders with
+ * generated bridge orders.
+ * @param nativeOrders Native orders.
+ * @param takerAmount Amount of taker asset to sell.
+ * @param opts Options object.
+ * @return Improved orders.
+ */
+export async function improveMarketSellAsync(
+    nativeOrders: OrderWithoutDomain[],
+    takerAmount: BigNumber,
+    opts?: Partial<ImproveOrdersOpts>,
+): Promise<OrderWithoutDomain[]> {
+    if (nativeOrders.length === 0) {
+        throw new Error(AggregationError.EmptyOrders);
+    }
+    const _opts = {
+        ...IMPROVE_ORDERS_OPTS_DEFAULTS,
+        ...opts,
+    };
+    const [orderInfos, dexQuotes] = await queryNetworkAsync(
+        nativeOrders,
+        getSampleAmounts(takerAmount, _opts.numSamples),
+        difference(SELL_SOURCES, _opts.excludedSources),
+        _opts.provider,
+    );
+    const nativePath = pruneDustFillsFromNativePath(
+        createSellPathFromNativeOrders(nativeOrders, orderInfos),
+        takerAmount,
+        _opts.dustThreshold,
+    );
+    const dexPaths = createPathsFromDexQuotes(dexQuotes, _opts.noConflicts);
+    const allPaths = [...dexPaths];
+    const allFills = flattenDexPaths(dexPaths);
+    if (!(_opts.excludedSources as ERC20BridgeSource[]).includes(ERC20BridgeSource.Native)) {
+        allPaths.splice(0, 0, nativePath);
+        allFills.splice(0, 0, ...nativePath);
+    }
+    const optimizer = new FillsOptimizer(_opts.runLimit);
+    const optimalPath = optimizer.optimize(
+        sortFillsByPrice(allFills),
+        takerAmount,
+        pickOptimalPath(allPaths, takerAmount),
+    );
+    if (!optimalPath) {
+        throw new Error(AggregationError.NoOptimalPath);
+    }
+    const [outputToken, inputToken] = getOrderTokens(nativeOrders[0]);
+    return createSellOrdersFromPath(inputToken, outputToken, simplifyPath(optimalPath), _opts.bridgeSlippage);
+}
+
+/**
+ * Improve a market buy operation by (potentially) merging native orders with
+ * generated bridge orders.
+ * @param nativeOrders Native orders.
+ * @param makerAmount Amount of maker asset to buy.
+ * @param opts Options object.
+ * @return Improved orders.
+ */
+export async function improveMarketBuyAsync(
+    nativeOrders: OrderWithoutDomain[],
+    makerAmount: BigNumber,
+    opts?: Partial<ImproveOrdersOpts>,
+): Promise<OrderWithoutDomain[]> {
+    if (nativeOrders.length === 0) {
+        throw new Error(AggregationError.EmptyOrders);
+    }
+    const _opts = {
+        ...IMPROVE_ORDERS_OPTS_DEFAULTS,
+        ...opts,
+    };
+    const [orderInfos, dexQuotes] = await queryNetworkAsync(
```

Comment:
Thoughts on using [`DevUtils.getOrderRelevantState`](https://github.com/0xProject/0x-monorepo/blob/development/contracts/dev-utils/contracts/src/OrderValidationUtils.sol#L51) in the Sampler and returning `[orderInfos, fillableTakerAssetAmount, isValidSignature]`.

We [perform this](https://github.com/0xProject/0x-monorepo/blob/development/packages/asset-swapper/src/utils/order_prune_utils.ts#L24) to ensure the order is valid and check how much is fillable. We could skip that check a by combining it in the Sampler and reduce an RPC call.

Reply:
Yeah, I'm down for something like that. It looks like there's some more complex pruning rules being applied. So maybe the sampler contract can query those states and pass them into an `orderFilter` callback in `opts`.

Reply:
Hmm, really all we care about is how much taker asset is fillable. I think I'll just have the contract return that quantity, which  will also be zero if the signature is invalid or if the order status is not `FILLABLE`.